### DISCUSSION TITLE
A2DP sound effect

### DISCUSSION THREAD

**Discussion by SandeepJagdev** [pschatzmann/arduino-audio-tools#469] [Q&A] [answered]
Hello i am working on ESP32A1S with AC101 codec. I want to implement ECHO or delay on bluetooth audio (A2DP) streaming (so i can make sound effect like large hall , 3D and echo on both bluetooth and line-in.

AudioKitStream kit; // Access I2S as stream
AudioEffects<GeneratorFromStream<effect_t>> effects(kit, 2 , 1.0); // apply effects on kit
GeneratedSoundStream<int16_t> in(effects);
StreamCopy copier(kit, in); // copy in to kit
Delay dly(1000,0.9,0.9,44100);

using this results no change on output.
i implement everything as per file "Audioeffects.h" but not any change in output
please help me. Thank you.

**Comment by pschatzmann** [comment]
You are not using the Delay anywhere: Please read the Wiki about Audio Effects or look at the examples

**Reply by SandeepJagdev** [reply]
effects.addEffect(dly);
  dly.setDuration(1000);
  dly.setDepth(0.5);
  dly.setFeedback(0.5);
  dly.setActive(1000>0);

Using this also no change in output

**Reply by pschatzmann** [reply]
what happens if you just use the provided default values ?
- leave constructor empty
- do not call setters

**Reply by SandeepJagdev** [reply]
Default values do not change any output. Please help.

**Reply by pschatzmann** [reply]
I might have some time later today to look into this

**Reply by SandeepJagdev** [reply]
please help

**Comment by pschatzmann** [comment]
I did some corrections and was debugging the result: I think it is doing now what it is supposed to do.
However I did not check the impact on the resulting audio.
The samples are stored in a ring buffer and the mixed together based on the depth. A depth of 0.5 gives half of the actual signal and half of the delayed sample

**Reply by SandeepJagdev** [reply]
The code you update in example works but when i done
AudioEffects<GeneratorFromStream<effect_t>> effects(kit,2,1.0); // apply effects on kit
and add delay . ESP32 restart again and again in never ending loop.

**Reply by pschatzmann** [reply]
Did you analyze the error message ? 

My  gut feeling is that your are just selecting a delay which is too long and you do not have enough RAM for this. In this case you could try to use PSRAM if your board has it.

**Reply by SandeepJagdev** [reply]
thanks a lot. but no difference in sound

**Reply by SandeepJagdev** [reply]
please help

**Comment by pschatzmann** [comment]
What execlty do you want to do with Delay ?
Why don't you use one of the STK based Delay implementations ?

**Reply by SandeepJagdev** [reply]
i want make echo and 3D effects on sound on both A2DP and line in

**Comment by pschatzmann** [comment]
There are plenty of possiblities if you use the STK based effects classes
Or you can generate your effect with Faust 
Or just add your own effect...

**Reply by SandeepJagdev** [reply]
Any example code or idea please ?? i am new in this field.

**Comment by pschatzmann** [comment]
Just look at the documentation: https://pschatzmann.github.io/arduino-audio-tools/html/classaudio__tools_1_1_audio_effect.html
You can have also a look at the [STK documentation](https://github.com/pschatzmann/arduino-stk) to get more details

There is also documentation about Faust in the Wiki

**Comment by pschatzmann** [comment, marked as answer]
After building a [proper infrastructure](https://www.pschatzmann.ch/home/2022/11/29/using-audiotools-in-jupyterlab/) to test the effects properly I have finally corrected the Delay to work properly. [Here are the tests](https://github.com/pschatzmann/arduino-audio-tools/blob/main/jupyter/JupyterGuitarEffects.ipynb)...

**Reply by john-** [reply]
That is really cool....and will be helpful once I start working with AudoTools!
